lib/srv/desktop/audit.go

```diff
@@ -145,8 +149,409 @@ func (s *WindowsService) onClipboardReceive(ctx context.Context, emitter events.
 	s.emit(ctx, emitter, event)
 }
 
+// onSharedDirectoryAnnounce adds the shared directory's name to the auditCache.
+func (s *WindowsService) onSharedDirectoryAnnounce(
+	ctx context.Context,
+	emitter events.Emitter,
+	id *tlsca.Identity,
+	sid string,
+	desktopAddr string,
+	m tdp.SharedDirectoryAnnounce,
+	tdpConn *tdp.Conn,
+) {
+	if err := s.auditCache.SetName(sessionID(sid), directoryID(m.DirectoryID), directoryName(m.Name)); err != nil {
+		// An error means the audit cache entry for this sid exceeded its maximum allowable size.
+		errMsg := err.Error()
+
+		// Close the connection as a security precaution.
+		if err := tdpConn.Close(); err != nil {
+			s.cfg.Log.Errorf("error when terminating sessionID(%v) for audit cache maximum size violation: %v", sid, err)
+		}
+
+		event := &events.DesktopSharedDirectoryStart{
+			Metadata: events.Metadata{
+				Type:        libevents.DesktopSharedDirectoryStartEvent,
+				Code:        libevents.DesktopSharedDirectoryStartFailureCode,
+				ClusterName: s.clusterName,
+				Time:        s.cfg.Clock.Now().UTC(),
+			},
+			UserMetadata: id.GetUserMetadata(),
+			SessionMetadata: events.SessionMetadata{
+				SessionID: sid,
+				WithMFA:   id.MFAVerified,
+			},
+			ConnectionMetadata: events.ConnectionMetadata{
+				LocalAddr:  id.ClientIP,
+				RemoteAddr: desktopAddr,
+				Protocol:   libevents.EventProtocolTDP,
+			},
+			Status: events.Status{
+				Success:     false,
+				Error:       errMsg,
+				UserMessage: "Teleport failed the request and terminated the directory sharing session as a security precaution",
```

Comment:
It didn't terminate the directory sharing session, it terminated the whole desktop session, right?

(same for below)

Reply:
9f820c2b9065b1968b2f71011767aece412fa672